Handle teams links

stuartleeks opened the pull request:
Decode embedded links from Microsoft Teams to display the underlying link and use it for applying defaults.

Closes #24

mortenn commented:
Thank you for the PR, seems like a good change.
Could you please correct whitespace to tabs before I merge it? :)

stuartleeks commented:
Apologies - I will fix that up!

stuartleeks commented:
Done. 

In the final commit on the PR I added an [editor config file and applied it](https://docs.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options?view=vs-2019). With the editor config you can set the style preferences and VS will use that to auto format etc to try to try to avoid issues in future. That said, I'm happy to either put that as a separate PR (and rebase this one on it to make the changes clearer) or drop that commit entirely if you prefer 😃

mortenn commented:
I am well aware of .editorconfig, but I have decided against using that in the past, as it has a terrible bug in that it locks you into indent size.
One of the big benefits to using tabs to indent is that each dev can themselves decide how big the indent should be, without forcing everyone to use the same tabwidth.
I, personally. prefer tw=2, but that might be a bit tight for some.

I've also had some interference between .editorconfig, VS and ReSharper, so I kind of abandoned introducing them for the source at my employer.

Though, as this is a very limited open source project, I suppose most my concerns with .editorconfig are fairly moot, though I am not entirely happy with forcing my preference of tw=2 on others, I can live with that consequence for this project, I think ^_^

mortenn commented on BrowserPicker/MainWindow.xaml.cs:
```diff
@@ -38,6 +38,7 @@ private void MainWindow_OnKeyUp(object sender, KeyEventArgs e)
 					case Key.D7: n = 7; break;
 					case Key.D8: n = 8; break;
 					case Key.D9: n = 9; break;
+					case Key.C: Clipboard.SetText(ViewModel.TargetURL); return;
```
sneaky bonus feature? ^_^

stuartleeks replied:
Haha - yes, it made life easier when working on the Teams links as it provided a quick way to copy the link :-)

stuartleeks commented:
I've just pushed a commit that adds support for the outlook link format. This required some changes to make the initialization async, but uses the approach outlined [here](https://github.com/mortenn/BrowserPicker/issues/24#issuecomment-760172571) of making the request and checking the `Location` header. This should make it easy to add support for other link shorteners :smiley:

stuartleeks commented:
And re editorconfig, let me know what you'd like me to do with it as far as this PR goes. Happy to take it out if you prefer :smiley:

mortenn commented on BrowserPicker/App.xaml.cs:
```diff
@@ -21,20 +24,69 @@ public App()
+			if (url.StartsWith("https://staticsint.teams.cdn.office.net/evergreen-assets/safelinks/"))
+			{
+				var uri = new Uri(url);
+				var queryString = uri.Query;
```
Would it not make sense to simply use `HttpUtility.ParseQueryString` here?

stuartleeks replied:
Oooh, good shout. Updated :-)

mortenn commented:
A further possible step here would be to add in logic that if the url path contains "authorize" then look for the value "redirect_uri" in the query string - that would handle oidc requests ^_^ it might be ok to always use redirect_uri, but if we make it too smart, it might fail unexpectedly on some urls.

Before I had a chance to look over your PR properly, I envisioned a solution where both the original URL and the underlying url is shown below eachother.

With async and my idea for looking up the location header, I was thinking the underlying url would not be displayed right away, but once the web request finished and returned a location header, if the domain of the underlying url differs from the original url, then display it - if they are the same, might just skip that

stuartleeks commented:
If the lookup is made async, how would the default rules work? Would you show the UI with the original URL but then hide it once the underlying URL is retrieved if it matches a rule?

mortenn commented:
Good question, I hadn't really thought that far ahead, just brainstorming here :)
I was thinking that remote calls should not be required to complete for UI to show, as that would slow down the user experience, especially if network is being slow.
Maybe ignore default rules for underlying URLs that only show up after doing a request?
Maybe make it a user option?
I'd tinker with it a bit and see what it feels like before I commit, were I working on the feature myself, probably.

stuartleeks commented:
Good point - I've pushed a commit that limits the lookup of the underlying URL to an amount in the config (defaults to 500ms).
For me, the real value of the lookup is to apply the default rules so I'd be happy to push the value to a higher amount in my config to get that :-)

mortenn commented:
I merged the other PRs, but this one now has a conflict, if you would care to resolve that, I'll merge this one too.
After this has been merged I'll need a few days to go over the code and see if I want to make any adjusments before I release a new build ^_^

stuartleeks commented:
Sure thing!

stuartleeks commented:
Rebased and fixed up the conflicts. I've got a local branch that builds on this and adds in support for url shorteners that return `Location` header. I can either merge those changes here or create as a separate PR after this one is merged (to keep the PRs smaller) - do you have a preference?

mortenn commented:
You can merge it here, it is kind of related after all.

stuartleeks commented:
Done :-)

mortenn commented:
When I open this project in Visual Studio with the .editorconfig, I am forced into having tw=4, even though my VS has it set to 2.
